dap: add SWD driver user API and Zephyr DAP Link header

jfischer-no opened the pull request:
Add SWD driver user API and Zephyr DAP Link header.

jfischer-no committed:
include: drivers: add SWD driver user API

Although the DAP subsystem is the only user in the tree, yet, we may have
other subsystems, libraries, or simple samples that could use this API
in the future.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dap: cmsis_dap: use SWD driver user API

Use SWD driver user API.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: dp: add common header for SWD drivers

Move the LUT to convert from simplified format to request packet
expected by the target, and the helper to calculate parity bits to
common header, so that they could be used by other drivers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dap: add header and macro to instantiate DAP context

Although we have the three layers, SWDP driver API, DAP stack, and USB
backend, to implement a debug probe, there is no way for the user to
instantiate the DAP context, which would allow it to be used with
different backends or to change the configuration at runtime.

The Debug Access Port (DAP) is an implementation of the ARM Debug
Interface (ADI) that is typically integrated into SoC.
With all three layers, we implement a link to the DAP on the SoC.
Let's call it Zephyr DAP Link and use the dap_link prefix. There is also
a MBED DAPLink project that implements similar functionality, but using
a different name would likely cause more confusion.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

fabiobaltieri commented on samples/subsys/dap/src/main.c:
```diff
@@ -12,27 +12,33 @@
 	struct usbd_context *sample_usbd;
 	int ret;
 
-	ret = dap_setup(swd_dev);
```
would it make sense to add a line in the migration guide mentioning that `dap_setup` needs to be replaced with `dap_link_init` + `dap_link_backend_usb_init`?

jfischer-no committed:
doc: migration: add note about DAP subsystem changes

Add note about DAP subsystem changes. Keep it general, as there will be
more changes.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
